contracts/Erdstall.sol

```diff
@@ -172,75 +177,64 @@ contract Erdstall {
         require(balance.epoch == frozenEpoch, "withdrawFrozen: wrong epoch");
         verifyBalance(balance, sig);
 
-        // Also recover deposits from broken epoch
-        uint256 value = balance.value + deposits[frozenEpoch+1][msg.sender];
+        // Also recover deposits from broken epochs
+        uint256 value = balance.value + frozenDeposits();
 
-        _withdrawFrozen(value);
+        _withdraw(frozenEpoch, value);
     }
 
-    function _withdrawFrozen(uint256 value) internal {
-        require(!frozenWithdrawals[msg.sender], "already withdrawn (frozen)");
-        frozenWithdrawals[msg.sender] = true;
+    // withdrawFrozenDeposit lets any user withdraw the deposits that they made
+    // in the challenged and broken epoch.
+    //
+    // This function should only be called by users who never received a balance
+    // proof and thus never had any balance in the system, to recover their
+    // deposits.
+    function withdrawFrozenDeposit() external {
+        ensureFrozen();
+
+        uint256 value = frozenDeposits();
+        require(value > 0, "no frozen deposit");
 
-        require(token.transfer(msg.sender, value), "withdrawFrozen: token transfer failed");
-        emit Withdrawn(frozenEpoch, msg.sender, value);
+        _withdraw(frozenEpoch, value);
     }
 
     // ensureFrozen ensures that the state of the contract is set to frozen if
-    // the last epoch has at least one unanswered challenge.
+    // the last epoch has at least one unanswered challenge. It is idempotent.
     //
-    // It is implicitly called by withdrawFrozen but can be called seperately if
-    // the contract should be frozen before anyone wants to withdraw.
+    // It is implicitly called by withdrawFrozen and withdrawFrozenDeposit but
+    // can be called seperately if the contract should be frozen before anyone
+    // wants to withdraw.
     function ensureFrozen() public {
         if (isFrozen()) { return; }
-        require(isLastEpochChallenged(), "no challenge in last epoch");
+        uint64 challEpoch = epoch() - 3;
```

Comment:
is this really safe?

Reply:
I guess we can assume that the challenger doesn't fail to call this in the right epoch. But if it is missed, looks like we'd fail to freeze.